qiita_db/portal.py

```diff
@@ -26,38 +150,191 @@ def get_studies_by_portal(portal):
             All study ids in the database that match the given portal
         """
         conn_handler = SQLConnectionHandler()
-        portal_id = convert_to_id(portal, 'portal_type', 'portal')
         sql = """SELECT study_id FROM qiita.study_portal
                  WHERE portal_type_id = %s"""
         return {x[0] for x in
-                conn_handler.execute_fetchall(sql, [portal_id])}
+                conn_handler.execute_fetchall(sql, [self._id])}
+
+    def _check_studies(self, studies):
+        conn_handler = SQLConnectionHandler()
+        # Check if any study IDs given do not exist.
+        sql = "SELECT study_id from qiita.study WHERE study_id IN %s"
+        existing = [x[0] for x in conn_handler.execute_fetchall(
+            sql, [tuple(studies)])]
+        if len(existing) != len(studies):
+            bad = map(str, (set(studies).difference(existing)))
+            raise QiitaDBError("The following studies do not exist: %s" %
+                               ", ".join(bad))
 
-    def add_study_to_portal(study, portal):
-        """Adds study to given portal
+    def add_studies(self, studies):
+        """Adds studies to given portal
 
         Parameters
         ----------
-        portal : str
-            Portal to associate with.
+        studies : list of int
+            Study ids to attach to portal
+
+        Raises
+        ------
+        QiitaDBError
+            Some studies given do not exist
         """
-        portal_id = convert_to_id(portal, 'portal_type', 'portal')
+        self._check_studies(studies)
+
+        conn_handler = SQLConnectionHandler()
+        # Clean list of studies down to ones not associated with portal already
+        sql = """SELECT study_id from qiita.study_portal
+                 WHERE portal_type_id != %s AND study_id IN %s"""
+        clean_studies = [x[0] for x in conn_handler.execute_fetchall(
+                         sql, [self._id, tuple(studies)])]
+
+        if len(clean_studies) != len(studies):
+            rem = map(str, set(studies).difference(clean_studies))
+            warnings.warn("The following studies area already part of %s: %s" %
```

Comment:
There's nothing specific about them in the documentation, but I would
suggest noting them and the reason why they will appear (we do the same
in load_template_to_dataframe ...). For reference here's the
documentation:
https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt

On (Jun-30-15|18:35), josenavas wrote:
> >          """
> > -        portal_id = convert_to_id(portal, 'portal_type', 'portal')
> > +        self._check_studies(studies)
> > +
> > +        conn_handler = SQLConnectionHandler()
> > +        # Clean list of studies down to ones not associated with portal already
> > +        sql = """SELECT study_id from qiita.study_portal
> > +                 WHERE portal_type_id != %s AND study_id IN %s"""
> > +        clean_studies = [x[0] for x in conn_handler.execute_fetchall(
> > +                         sql, [self._id, tuple(studies)])]
> > +
> > +        if len(clean_studies) != len(studies):
> > +            rem = map(str, set(studies).difference(clean_studies))
> > +            warnings.warn("The following studies area already part of %s: %s" %
> 
> Can you use the QiitaDBWarnings?
> 
> @ElDeveloper do the warnings need to be in the numpy doc?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiita/pull/1316/files#r33641712